kaggle_environments/envs/lux_ai_2022/luxai2022/env.py

```diff
@@ -0,0 +1,618 @@
+                        for u in units:
+                            if u.unit_id != surviving_unit.unit_id:
+                                destroyed_units.add(u)
+                        new_units_map_after_collision[pos_hash].append(surviving_unit)
+            self.state.board.units_map = new_units_map_after_collision
+
+            for u in destroyed_units:
+                self.destroy_unit(u)
+
+            # for recharging actions, check if unit has enough power. If not, add action back to the queue
+            
+            for unit, recharge_action in actions_by_type["recharge"]:
+                recharge_action: RechargeAction
+                if unit.power < recharge_action.power:
+                    unit.action_queue.insert(0, recharge_action)
+                    # by default actions with repeat=True will be placed to the back of queue
+                    # remove from back of queue if we re-inserted into the front
+                    if recharge_action.repeat:
+                        unit.action_queue = unit.action_queue[:-1]
+
+
+            for factory, factory_water_action in actions_by_type["factory_water"]:
+                water_cost = factory.water_cost(self.env_cfg)
+                factory.cargo.water -= water_cost # earlier validation ensures this is always possible.
+                indexable_positions = ([v[1] for v in factory.grow_lichen_positions], [v[0] for v in factory.grow_lichen_positions])
+                self.state.board.lichen[indexable_positions] += 2
+                self.state.board.lichen_strains[indexable_positions] = factory.num_id
+            
+            self.state.board.lichen -= 1
+            self.state.board.lichen = self.state.board.lichen.clip(0)
+            self.state.board.lichen_strains[self.state.board.lichen == 0] = -1
+
+            # resources refining
+            for agent in self.agents:
+                factories_to_destroy: Set[Factory] = set()
+                for factory in self.state.factories[agent].values():
+                    factory.refine_step(self.env_cfg)
+                    factory.cargo.water -= self.env_cfg.FACTORY_WATER_CONSUMPTION
+                    if factory.cargo.water < 0:
+                        factories_to_destroy.add(factory)
+                for factory in factories_to_destroy:
+                    # destroy factories that ran out of water
+                    self.destroy_factory(factory)
+            # power gain
+            if is_day(self.env_cfg, self.state.real_env_steps):
+                for agent in self.agents:
+                    for u in self.state.units[agent].values():
+                        u.power = u.power + math.ceil(self.env_cfg.ROBOTS[u.unit_type.name].CHARGE * weather_cfg["power_gain_factor"])
+            for agent in self.agents:
+                for f in self.state.factories[agent].values():
+                    f.power = f.power + math.ceil(self.env_cfg.FACTORY_CHARGE * weather_cfg["power_gain_factor"])
```

Comment:
aren't factories immune to weather?

Reply:
yeah let's just make them immune, they don't use solar panels